Sources/ViewType.swift

```diff
@@ -0,0 +1,225 @@
+//
+//  Copyright © 2020 Rosberry. All rights reserved.
+//
+
+import UIKit
+
+private var layoutAssociationKey = "layout_association_key"
+
+public enum ViewType {
+    case view(UIView)
+    case layer(CALayer)
+}
+
+extension ViewType {
+
+    var layout: Layout {
+        if let value = objc_getAssociatedObject(self, &layoutAssociationKey) as? Layout {
+            return value
+        }
+        else {
+            let layout: Layout
+            switch self {
+            case .view(let view):
+                layout = UIViewLayout(view: view)
+            case .layer(let layer):
+                layout = CALayerLayout(layer: layer)
+            }
+            objc_setAssociatedObject(self, &layoutAssociationKey, layout, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+            return layout
+        }
+    }
```

Comment:
That's why you shouldn't use enums when you need to have a stored object.

Reply:
Even without enum i need associated object to store some instance that provides access to layout properties

Reply:
Well, no, you don't need an associated object at all. You can do something like this:
```
public class ElementType {
    static func view(_ view: UIView) -> ElementType {
        return ViewType(view: view)
    }

    static func layer(_ layer: CALayer) -> ElementType {
        return LayerType(layer: layer)
    }

    let layout: Layout
    let layer: CALayer

    init(layer: CALayer, layout: Layout) {
        self.layer = layer
        self.layout = layout
    }
}

public class ViewType: ElementType {
    public let view: UIView

    init(view: UIView) {
        self.view = view
        let layout = ViewLayout(view: view)
        super.init(layer: view.layer, layout: layout)
    }
}

public class LayerType: ElementType {
    init(layer: CALayer) {
        let layout = LayerLayout(layer: layer)
        super.init(layer: layer, layout: layout)
    }
}
```
Switched naming here a bit, but you get the idea. 
You're losing convenient `switch` by getting rid of enum, but the best part is that you won't need to use `switch` at all, because you'll be able to access layers directly.